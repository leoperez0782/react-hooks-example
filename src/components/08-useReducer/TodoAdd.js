import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({handleAddTodo}) => {
    const [{description}, handleInputChange, reset] = useForm({
        description: ''
    });

    const handleSubmit =(e) =>{
        e.preventDefault();

        if(description.trim().length <=1){
            return;
        }
        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        };
        
        handleAddTodo( newTodo );//manda la accion al reducer.
        reset();
    }
    return (
        <>
        <h4>Agregar TODO</h4>
                    <hr></hr>
                    <form onSubmit={handleSubmit}>
                        <input
                            type="text"
                            name="description"
                            placeholder="Aprender..."
                            autoComplete="off"
                            className="form-control"
                            value={description}
                            onChange={handleInputChange}
                        />
                        <div className="d-grid gap-2">
                            <button
                                className="btn btn-outline-primary mt-1 "
                                type="submit">
                                Agregar
                            </button>
                        </div>

                    </form> 
        </>
    )
}
