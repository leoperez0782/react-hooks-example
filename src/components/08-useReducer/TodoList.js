import React from 'react'
import { TodoItem } from './TodoItem'

export const TodoList = ({ todos, handleDelete, handleToggle }) => {
    return (
        <ul className="list-group list-group-flush">
            {
                todos.map((todo, i) => (
                    <TodoItem
                        todo={todo}
                        key={todo.id}
                        index={i}
                        handleDelete={() => handleDelete(todo.id)}
                        handleToggle={() => handleToggle(todo.id)}
                    />
                ))
            }
        </ul>
    )
}
