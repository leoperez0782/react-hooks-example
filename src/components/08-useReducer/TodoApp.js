import React, { useEffect, useReducer } from 'react'
import { todoReducer } from './todoReducer';
import './styles.css';

import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';
export const TodoApp = () => {

    const init = ()=>{
       return JSON.parse(localStorage.getItem('todos')) || [];
    }

   
    const [todos, dispatch] = useReducer(todoReducer, [], init);

    

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
     }, [todos]);
 
     const handleToggle = (todoId)=>{
         dispatch({
             type:'toggle',
             payload: todoId
         });
     }

     const handleAddTodo =(newTodo) =>{
         dispatch({
            type: 'add',
            payload: newTodo
        });
     }
     const handleDelete =(todoId) =>{
        //crear la accion
        const action ={
            type: 'delete',
            payload: todoId
        }
        dispatch(action);
     }
    
    return (
        <>
            <h1>TodoApp ({todos.length})</h1>
            <hr></hr>
            <div className="row">
                <div className="col-7">
                
                    <TodoList todos={todos} 
                    handleDelete={handleDelete} 
                    handleToggle ={handleToggle} />
                </div>
                <div className="col-5">
                    <TodoAdd 
                    handleAddTodo ={handleAddTodo}/>
                </div>

            </div>
        </>
    )
}
