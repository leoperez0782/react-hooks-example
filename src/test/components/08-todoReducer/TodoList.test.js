import React from 'react';
import { shallow } from 'enzyme';
import { TodoList } from '../../../components/08-useReducer/TodoList';
import { demosTodos } from '../../fixtures/demosTodos';



describe('Pruebas en <TodoList />', () => {

    const handleDelete = jest.fn();
    const handleToggle = jest.fn();

    const wrapper = shallow(
        <TodoList
        todos={demosTodos}
        handleDelete = {handleDelete}
        handleToggle = {handleToggle}
        />
    );
    
    test('debe de mostrarse correctamente', () => {
        
        expect( wrapper ).toMatchSnapshot();
    });

    test('debe de tener 2 <TodoItem/>', () => {
        
        expect( wrapper.find('TodoItem').length ).toBe( demosTodos.length );
        expect( wrapper.find('TodoItem').at(0).prop('handleDelete') ).toEqual( expect.any(Function));

    })
    
});
