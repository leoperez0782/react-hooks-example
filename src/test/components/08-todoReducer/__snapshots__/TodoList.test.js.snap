// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoList /> debe de mostrarse correctamente 1`] = `
<ul
  className="list-group list-group-flush"
>
  <TodoItem
    handleDelete={[Function]}
    handleToggle={[Function]}
    index={0}
    key="1"
    todo={
      Object {
        "desc": "Aprender react",
        "done": false,
        "id": 1,
      }
    }
  />
  <TodoItem
    handleDelete={[Function]}
    handleToggle={[Function]}
    index={1}
    key="2"
    todo={
      Object {
        "desc": "Aprender mongo",
        "done": false,
        "id": 2,
      }
    }
  />
</ul>
`;
